### ISSUE TITLE
ml.dtw randomly freezes Pd when [train( is sent

### ISSUE THREAD

**Issue by graphographer** [cmuartfab/ml-lib#143]
Using Pd 0.46.7 as well as 48.0-1 on Ubuntu 16.04 LTS.

**Comment by graphographer** [comment]
I thought I had found some patterns related to sending [clear( before [train(, or opening and closing and reopening the patch, but I now think the freezing is random.

**Comment by jamiebullock** [comment]
Oh cool, you’re using the Linux builds... those are relatively new.

Is it freezing only with ml.dtw or all externals?

Also, are you familiar with debugging with the gdb command line tool? If so, I can send you a debug build and there are a few commands we can run to debug the issue.

**Comment by graphographer** [comment]
I'll have limited time in the next few weeks, but I'm happy to try to go through all the objects and quickly test them. Just in the past five minutes, ml.anbc seemed OK in this regard, whereas ml.dtree can't even load before giving a segmentation fault and crashing Pd entirely.

I've only had time to focus on ml.dtw, which is my main interest. It's for accelerometer/gyroscope-based gesture recognition for a music instrument.

I'm happy to learn the rudiments of GDB and report back.

**Comment by graphographer** [comment]
Hi there! Checking in to see if there's anything I can do to help with this.

**Comment by jamiebullock** [comment]
Hi @graphographer, sorry for the delay.

Would you mind trying the new Linux build, which can be found [here](https://github.com/cmuartfab/ml-lib/releases/tag/0.18.0) and letting me know if it solves your crashing issues.

I can't make it crash here, so if it does crash, can you send me a Pd patch that will cause it to crash on your machine?

Thanks!

**Comment by graphographer** [comment]
Certainly! Thanks.

**Comment by graphographer** [comment]
Hi there, sorry for the long delay. I was finally able to try the linux builds you posted here, and the issue seems to persist. I have attached here the patch I've been using to test ml_dtw. Perhaps the patch and my explanation of it will turn up any misapprehensions on my part about how the library works! On the other hand, it seems to me that the lib still shouldn't cause Pd to crash/freeze under any normal circumstance, no matter how bizarre the data input may be.

In any case, here is how I use the patch. I click the toggle which initiates record mode. I then pull the horizontal slider up, then down (like a triangle wave basically), and click the toggle again to stop the recording. I do this maybe three times for class 1. Then I switch the radio button to class 2, and repeat the procedure, but starting with the slider at its max, and going down then back up. Then I click "train" whereupon Pd becomes unresponsive.

I have varied the procedure by perhaps only adding one sample set to one class, then clicking "train" which sometimes seems to results in success, but does inevitably seem to cause Pd to crash if I repeat this process.

[dtw_test.pd.txt](https://github.com/cmuartfab/ml-lib/files/1935700/dtw_test.pd.txt)
